semantics/src/lib.rs

```diff
@@ -59,9 +59,58 @@ pub struct ContractAttributes {
 pub struct ExpressionAttributes {
     pub typ: Type,
     pub location: Location,
+    pub move_to: Option<Location>,
 }
 
 impl ExpressionAttributes {
+    pub fn new(typ: Type, location: Location) -> Self {
+        Self {
+            typ,
+            location,
+            move_to: None,
+        }
+    }
+
+    /// Creates a new `ExpressionAttributes` from the parameter with a move
+    /// to the default location for the type, if needed.
+    pub fn with_default_move(mut self) -> Result<Self, SemanticError> {
+        let move_to = match self.typ {
+            Type::Base(_) => Location::Value,
+            Type::Array(_) => Location::Memory,
+            Type::Tuple(_) => Location::Memory,
+            Type::String(_) => Location::Memory,
+            Type::Map(_) => return Err(SemanticError::CannotMove),
+        };
+
+        if self.location != move_to {
+            self.move_to = Some(move_to)
+        }
+
+        Ok(self)
+    }
+
+    pub fn with_value_move(mut self) -> Result<Self, SemanticError> {
+        match self.typ {
+            Type::Base(_) => {}
+            _ => return Err(SemanticError::CannotMove),
+        }
+
+        if self.location != Location::Value {
+            self.move_to = Some(Location::Value);
+        }
+
+        Ok(self)
+    }
+
+    /// The final location of an expression after a move takes place.
+    pub fn final_location(&self) -> Location {
+        if let Some(location) = self.move_to.clone() {
+            return location;
+        }
+
+        self.location.clone()
+    }
+
     /// Convenience method for matching type and location.
     pub fn to_tuple(&self) -> (Type, Location) {
```

Comment:
rename this to `type_and_location` and think about whether or not this should be the `final_location` or initial `location`